### PR TITLE
Added node for Navigation links

### PR DESCRIPTION
**Pull request by DarkKilauea**
Addresses godotengine/godot-proposals#2527

Test Project: https://github.com/DarkKilauea/godot-nav-link-demo

Things to do:

- [x] Add 2D version of navigation link node.
- [x] Add 2D editor plugin for navigation link nodes.
- [x] Add support for navigation links to the 2D navigation server.
- [x] Create debug visualization for `NavigationLink3D`.
- [x] Create debug visualization for `NavigationLink2D`.
- [x] Fix `get_path` not always going through the link's points (optimization issue?).
- [x] Rework travel cost to act as a multiplier, like `NavigationRegion`.
- [x] Support un-optimized paths (currently broken due to missing edge logic).

How it works:

Navigation links allow you to inform the navigation system about routes that don't involve traveling the surface of the navigation mesh.  This can include things like jumping off surfaces, teleporters, rappelling points, or other game specific activities.

Links do not handle any of the game logic, animations, or other aspects of taking an alternative route, they only inform the system that those routes exist so `get_path` can consider them.  This enables scenarios that currently require working around the navigation server today.

How to use:

1. Create a `NavigationLink3D` node.
2. Place it in your scene.
3. Move the start point to where you want the link to begin.
4. Move the end point to where the want the link to lead to.

Screenshot:
nav_link_editor

Video showing it in a test scene:
nav_link_demo

### CODE REVIEW

**Comment by venilark** [comment]
when the Agent is close to the navlink, will it receive a signal or something so we know it is time to execute our game logic (like start climbing the ladder or jumping)?

**Comment by DarkKilauea** [comment]
> when the Agent is close to the navlink, will it receive a signal or something so we know it is time to execute our game logic (like start climbing the ladder or jumping)?

That isn't currently planned for this PR.  Signaling would require `get_path` to provide more information than just the vertices of the path and `NavigationAgent` to pass that information in its signals.  I think some discussion is needed on how exactly that should be done.

In the meantime, `Area` can be used to trigger game logic as a physics body enters it, allowing for a similar effect to be implemented.  I plan to demo that in the test project before this PR is complete.

**Comment by venilark** [comment]
> > when the Agent is close to the navlink, will it receive a signal or something so we know it is time to execute our game logic (like start climbing the ladder or jumping)?
> 
> That isn't currently planned for this PR. Signaling would require `get_path` to provide more information than just the vertices of the path and `NavigationAgent` to pass that information in its signals. I think some discussion is needed on how exactly that should be done.
> 
> In the meantime, `Area` can be used to trigger game logic as a physics body enters it, allowing for a similar effect to be implemented. I plan to demo that in the test project before this PR is complete.

For demo purposes it's fine but for the final implementation we need a signal or some other way to know the Agent really has to use whatever shortcut is inside the Area, otherwise it will always use it as soon as it gets triggered, like when the target and the area are lined up and the agent only needs to walk in a straight line to the target.

**Comment by smix8** [comment]
What this pr / draft basically adds is the option to merge navigationmesh edges over arbitrary (long) distances  by using a scene node markup for the default pathfinding.

It would be a good feature to have to solve some situations. A simple "please try to jump down here instead of walking around" markup for pathfinding is all that is required for some smaller games made by the average Godot user. It would be quite user friendly and especially not overload newer users with way to much options that require additional setup.

I would focus on the 2D part and Editor usablity for this pr. Additional features like radius or chaining links ... I have my doubt over their usefulness or how much work it would involve to make and keep them bugfree with the current pathfinding and would maybe revisit them later.

I think as long as this pr sticks to the scope of connecting navmesh edges it can comfortably co-exists, if not complement, the more complex solution required for the proposal and would be a valuable addition to Godot.

This pr would not be able to solve the proposal even if it wanted to. Working with just the navmesh edges and the default navigation pathfinding is too limiting. It does not help with the majority of gameplay cases or provide any solution to the performance concerns from the linked proposal. In fact as a tradeoff for using these links it would worsen the default pathfinding performance with each added link for each single path query.

**Comment by DarkKilauea** [comment]
I have updated my approach to no longer be based on polygon edges, now links join polygons together, providing an alternative route than going through the edge.

I agree with your priorities @smix8 and I'll focus on taking care of those areas first.  I went ahead and removed radius support and chaining links together from my TODO list, we can address those later.

As for fulfilling requirements, looking at the original list from the proposal, I believe most of them are being addressed:

1. Allow locked un-navigable areas that can be later unlocked without re-baking nav-mesh.
    - This can be done by linking two regions together and controlling access via the `enabled` property or via `navigation_layers`.
2. Prevent specific agents from using specific navigation paths, or preventing some agents from accessing an area altogether.
    - This can be done via `navigation_layers`.
3. Allow verticality on baked nav-meshes.
    - This can be done by creating a link between a higher portion of the nav-mesh to a lower portion.
4. Allow prioritization or avoidance of specific paths to some extent.
    - Will be controllable with `travel_cost` and `enter_cost`.
5. Enable detecting going through a door or ladder or similar easier, without resorting to hacks.
    - This one is not currently planned, since I believe it requires changes to the API of `get_path`.
 
I'm debating adding a `bi-directional` property to match what Unreal and Unity have, especially since chaining is off the table for the moment.

Are there other requirements you think are important for this first version that I haven't addressed here?

As for performance, I believe the bulk of the cost is during link generation for the nav-mesh as a whole.  Once the links are created, they should only be considered during `get_path` if a polygon (which is being considered) has a link leading out from it.  The approach is pretty similar to how the existing connections out from edges are working now.

**Comment by smix8** [comment]
I think adding a bidirectional property would be good for this pr as this would be something very commonly used and placing 2 links on top of each other to cover both directions would create awkward usability in the Editor Viewport.

The performance concerns can be left outside of this pr. The current A*Star pathfinding algo cannot deal with those link situations efficiently as the current A*Star only works well in situations with non-weighted and connected cells. This needs to be adressed with some kind of preprocessing and waypoint / section caching at some point.

**Comment by smix8** [comment]
@DarkKilauea
Some Navigation internals were changed recently with pr https://github.com/godotengine/godot/pull/63584.
You might want to adapt your branch to these changes.

**Comment by DarkKilauea** [comment]
Thanks for the heads up @smix8, I've rebased and removed all usages of `std::vector`.

Other updates:
- `NavigationServer2D` now has bindings for all navigation link functions.
- Removed logic around start and end position radius and replaced it with a search radius project setting.  Links now bind to a single polygon for the start and a single polygon for the end, based on whatever polygon is closest.
- Added support for `bi_directional`.
- Added support for both `travel_cost` and `enter_cost`, which work the same way as their `NavigationRegion` counterparts.  Enter cost will be "paid" when using the link as a constant cost.  Travel cost is multiplied by the length between the start and end points.

**Comment by Calinou** [comment]
> Added support for bi_directional.

We have it spelled `bidirectional` elsewhere, so I'd rename `bi_directional` to `bidirectional`.

**Comment by DarkKilauea** [comment]
> > * Added support for `bi_directional`.
> 
> We have it spelled `bidirectional` elsewhere, so I'd rename `bi_directional` to `bidirectional`.

Fixed.

**Comment by DarkKilauea** [comment]
Update:

I think this is ready for review now.  I've added full 2D support and updated the debug visualizations for both 2D and 3D.  I've also written documentation for all the new additions and fix the last remaining bugs I'm aware of.

I did end up rewriting my approach processing the links in the navigation map itself.  Instead of treating the links as a separate entity, instead a synthetic polygon is created which creates a "bridge" between polygons.  I got the idea from Detour and it seems to be playing nicer with how `get_path` is written.

Please test out the feature and let me know if you run into any odd behavior that needs to be fixed.

**Comment by DarkKilauea** [comment]
Rebased on top of current master and incorporated changes made to debug in NavigationServer2D and fixes made to the NavigationRegion3D gizmo.

**Comment by YuriSizov** [comment]
cc @smix8 can you please evaluate this?

**Comment by smix8** [comment]
Needs more testing as I get random engine crashes with this pr after a few seconds. Seems the while loop of the NavMap::get_path() function gets stuck in an endless loop when navlinks are used to connect navigationregion islands which causes the crash.

**Comment by DarkKilauea** [comment]
@smix8 I'm not able to reproduce any crashes, even after creating a stress test with 5000 agents (based on the conversation we had in the dev chat).

image

Would you be able to upload your project that is crashing or create a minimal project that reproduces the issue?

**Comment by DarkKilauea** [comment]
@smix8 I found a bug where points would be added to the synthetic polygon created by the link every time the map was updated, leading to large memory consumption over time.

Please take another look at the PR and see if it fixes your crashing issue.

**Comment by smix8** [comment]
@DarkKilauea
Yes this fixes the crashes. The demo now runs stable again with 10000+ actors without a single crash.

**Comment by smix8** [comment]
I didn't encounter any more bugs (so far). We wouldn't get any of the required user feedback to fix the gameplay and Editor issues while this is not merged so from my part this is good to merge for alpha.

I would like to ask @AndreaCatania @groud @Scony @lawnjelly to take a quick look at the changes in `nav_map.cpp` if they have time cause they contributed that part originally or are far more seasoned with C++ then I am. The entire pr looks large but a lot of it is just required copy&paste from the navigation regions so I don't want to bother them with reviewing all that.

Usability should be improved with follow up pr's e.g. the Editor 3D gizmos parts to place the positions are really fiddly to use or sometimes very hard to see on more complex scenes. Would be nice to have a "sticky" option for the Editor that aligns the link with the closest collision ground position or closest navmesh position but that should not be part of this pr.

**Comment by DarkKilauea** [comment]
Agreed.  Here are my future plans for this feature:

1. Experiment with changing `NavigationLink*` into a node per endpoint of the link, instead of representing the entire link as a whole.  Doing so would enable the built in support for "Snap to Floor" and allow making the endpoints children of other nodes so they follow them around.  This is not currently possible with gizmo handles.
2. Expose information about whether the agent is traveling a normal navigation polygon or a link when a path is requested.  This is currently blocked on #62429.
3. Update `NavigationAgent*` to fire a signal (or update an existing signal to do so) when agents enter a link.  This would allow devs to implement game specific logic to handle navigation of the link.  I believe this is the last major request of the proposal that sparked this feature.

**Comment by venilark** [comment]
> Agreed. Here are my future plans for this feature:
> 
>     1. Experiment with changing `NavigationLink*` into a node per endpoint of the link, instead of representing the entire link as a whole.  Doing so would enable the built in support for "Snap to Floor" and allow making the endpoints children of other nodes so they follow them around.  This is not currently possible with gizmo handles.
> 
>     2. Expose information about whether the agent is traveling a normal navigation polygon or a link when a path is requested.  This is currently blocked on [Add NavigationPathQuery objects and NavigationServer query_path() #62429](https://github.com/godotengine/godot/pull/62429).
> 
>     3. Update `NavigationAgent*` to fire a signal (or update an existing signal to do so) when agents enter a link.  This would allow devs to implement game specific logic to handle navigation of the link.  I believe this is the last major request of the proposal that sparked this feature.

last one sounds excelent :)
anyway, are these changes planned to be in 4.0 release or 4.1 (so in the very distant future)?

**Comment by DarkKilauea** [comment]
@venilark I would like to get them into 4.0 if possible, but that is mainly up to the Godot maintainers at this point.  

Signaling when a link is reached also depends on #62429, since I need send back more information than `map_get_path` is able to do.  I'd likely implement it by annotating each of the spans of the computed path with whether they are a link and which link they correspond to.  Then, when `NavigationAgent` reaches those spans, it can signal with the correct link.

**Comment by DarkKilauea** [comment]
Rebased, updated PR to use TypedArray<RID> for `get_links` and using the new `EditorUndoRedoManager`.

**Comment by DarkKilauea** [comment]
Rebased and adjusted for math_func renames.

File: modules/navigation/nav_link.h
Comment (akien-mga) [inline modules/navigation/nav_link.h:41]: Vector3 are default initialized so you don't need to initialized them explicitly (might apply to more files / other vector types, I just saw this while browsing idly).

**Review by akien-mga — APPROVED** [review]
Should be good to merge as a first step if nobody objects. Would still be nice to get some more input from navigation minded people mentioned earlier :)

**Comment by lawnjelly** [comment]
I will try and have a quick glance through too this evening. Had been putting it off as had a lot of other work and it is 2.5K lines of code. :grin:

File: modules/navigation/nav_utils.h
Comment (lawnjelly) [inline modules/navigation/nav_utils.h:125]: What was reasoning for changing this to int out of interest?

File: modules/navigation/nav_region.h
Comment (lawnjelly) [inline modules/navigation/nav_region.h:43]: Where do these enter and travel costs for the region go to in the new version? Or are they removed?

File: modules/navigation/nav_utils.h
Comment (lawnjelly) [inline modules/navigation/nav_utils.h:98]: Should this be `Vector` or `LocalVector`? (Vector is more heavyweight, but can be bound for gdscript)

**Comment by akien-mga** [comment]
Thanks!

---

And sorry, I forgot to check @lawnjelly's outstanding review comments before merging (was testing a new workflow to merge multiple PRs at once to reduce the CI congestion).

I guess these will have to be handled in a follow-up PR :)

File: modules/navigation/nav_utils.h
Comment (DarkKilauea) [reply modules/navigation/nav_utils.h:125]: I'm using `-1` to signal to `get_path` that this polygon is part of a link and the midpoints of the edge should not be added to the path.  Instead, only the end points are.

File: modules/navigation/nav_utils.h
Comment (DarkKilauea) [reply modules/navigation/nav_utils.h:98]: I'm not sure, so I left it alone here.  However, I don't think we ever use the COW functionality, so it may make sense to change it.

File: modules/navigation/nav_region.h
Comment (DarkKilauea) [reply modules/navigation/nav_region.h:43]: They were moved to `NavBase`, a common base class that represents properties used as part of calculating the path in `get_path`.  Both `NavRegion` and `NavLink` share this base class since they act similarly.

It also removed the need to detect the type of `NavRegion` or `NavLink` in the `get_path` logic.

File: modules/navigation/nav_link.h
Comment (DarkKilauea) [reply modules/navigation/nav_link.h:41]: I can fix this in a future PR.

**Comment by venilark** [comment]
Did you decide in the PR meeting if this feature and the upcoming improvements/new PRs will be part of 4.0 release?

**Comment by akien-mga** [comment]
It is merged in the `master` branch so it's already part of 4.0.

**Comment by Riordan-DC** [comment]
Wow, amazing work. A day ago I found out I needed this feature and now it is here!
I wonder if this change can be shared with us on the 3.x branch?

**Comment by venilark** [comment]
> It is merged in the `master` branch so it's already part of 4.0.

Amazing!
Kudos to all of you, can't wait to see all the enemies and NPCs jumping and chasing us all over the map, this feature will for sure be a massive leap forward regarding agent navigation.
Just as a suggestion, if not for a new PR then for some future plugin (like the one @smix8 made), an autojump option that uses the connection between both links as a path for the jump (plus a signal to trigger the jump animation), basically what Unity does, except here it would be optional so the user can choose between their own implementation if needed or the basic, simple jump right out the gate.

**Comment by Frontrider** [comment]
Yay! Thank you!

**Comment by DarkKilauea** [comment]
Updated the image in the initial description and added a short video showing the feature in action.

I'd like to port this to 3.x eventually, though that isn't at the top of my priority list at the moment, so if someone else beats me to it, that's fine. :)

**Pull request merged**
